segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,457 @@
+                       () -> !closed.get(),
+                       () -> getDelayedFutureIfNeeded().thenComposeAsync(v -> processEvents(), executor),
+                   executor)
+                       .handle((r, ex) -> {
+                           if (ex != null) {
+                               log.warn("{}: Terminated due to unexpected exception.", this.traceObjectId, ex);
+                           } else {
+                               log.info("{}: Terminated.", this.traceObjectId);
+                           }
+                           return null;
+                       });
+        }
+
+        /**
+         * Gets a delayed future when either i) an error has been experienced in the last iteration, or ii) if there is
+         * no outstanding data to process. Otherwise, we do not delay the next iteration and return a completed future.
+         *
+         * @return A future that may delay the next processing iteration if there has been an error in the last
+         * processing iteration or there is no outstanding data to process.
+         */
+        private CompletableFuture<Void> getDelayedFutureIfNeeded() {
+            return failedIteration.get() || outstandingBytes.get() == 0 ? Futures.delayedFuture(iterationDelay, executor) :
+                    Futures.delayedFuture(Duration.ZERO, executor);
+        }
+
+        /**
+         * Returns a {@link CompletableFuture} that results from the execution of the following tasks for each EventProcessor:
+         * i) Read available data in the Segment, ii) deserialize at most {@link EventProcessorConfig#getMaxItemsAtOnce()},
+         * iii) execute the handler function for the {@link EventProcessor}, and iv) truncate the internal Segment.
+         */
+        private CompletableFuture<Void> processEvents() {
+            final Timer iterationTime = new Timer();
+            return readEvents()
+                    .thenComposeAsync(this::applyProcessorHandler, executor)
+                    .thenComposeAsync(readResult -> truncateInternalSegment(readResult, iterationTime), executor)
+                    .handleAsync((r, ex) -> {
+                        // If we got an exception different from NoDataAvailableException, report it as something is off.
+                        if (ex != null && !(Exceptions.unwrap(ex) instanceof NoDataAvailableException)) {
+                            log.warn("{}: Processing iteration failed, retrying.", this.traceObjectId, ex);
+                            failedIteration.set(true);
+                        } else {
+                            failedIteration.set(false);
+                        }
+                        return null;
+                    });
+        }
+
+        private CompletableFuture<EventsReadAndTruncationPoints> readEvents() {
+            AtomicLong startOffset = new AtomicLong(0L);
+            AtomicInteger readLength = new AtomicInteger(0);
+            return CompletableFuture.supplyAsync(() -> {
```

Comment:
I don't think we need CompletableFuture.supplyAsync() here.
simply call 
```
startOffset.set(segment.getInfo().getStartOffset());
readLength.set((int) Math.min(segment.getInfo().getLength() - startOffset.get(), MAX_TOTAL_EVENT_SIZE));
rr = segment.read(startOffset.get(), readLength.get(), containerOperationTimeout);;
return AsyncReadResultProcessor.processAll(rr, executor, containerOperationTimeout), executor)
                                    .thenApply(inputData -> deserializeEvents(inputData, startOffset.get(), readLength.get()));
```

Reply:
I have already answered that one, if I remember correctly. The reason is the following: with the above code, if an error happens in the `segment.read()` call, the exception is thrown and handled by the `handle` method following `Futures.loop()`, so the process terminates. As we want _any_ error that happens at any stage of the `Futures.loop()` to be handled by the `handle` method of `processEvents()`, we need to make that read step yet another future in the chain that ends in that `handle` method.